src/compression/mod.rs

```diff
@@ -510,74 +521,98 @@ mod optimize_bytes {
 
     /// Interleave the bytes such that the second half of the array is every other byte.
     pub fn interleave_byte_blocks(separated: &mut [u8]) {
-        // TODO without extra allocation!
-        let mut interleaved = vec![0; separated.len()];
-        let (first_half, second_half) = separated
-            .split_at((separated.len() + 1) / 2);
+        SCRATCH_SPACE.with(|scratch_space| {
+            // reuse a buffer if we've already initialized one
+            let mut buffer = scratch_space.take();
+            if buffer.len() < separated.len() {
+                // efficiently create a zeroed Vec by requesting zeroed memory from the OS
+                buffer = vec![0u8; separated.len()];
+            }
```

Comment:
Let's refactor this part into it's own function next to the declaration of the static variable. Not only is it roughly the same for both usages, but more importantly, it will separate the memory management from the two actual byte algorithms. This will simplify both algorithms. What do you think? 

Of course, only if we don't use `lazy_static` or if using it still requries some memory management.


Note: Including the finalizing statements of course
```
            // write out the results
            separated.copy_from_slice(&interleaved);
            // save the internal buffer for reuse
            scratch_space.set(buffer);
```

Reply:
Yes, it is faster. The alternative is a `memcpy()` for copying over old data plus `memset()` to expand the buffer. We don't need the old data, and requesting zeroed memory from the OS is faster than a `memset()`.

Reply:
But I understand this is done once per image. So it doesn't really matter if this is fast or not since it's done so rarely.

Reply:
I've factored out allocation reuse into its own function.

Reply:
great work :)